core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1647,24 +1652,29 @@ void rollbackOrProcessStateUpdates(
             }
 
             writeShareGroupState(stateBatches).whenComplete((result, exception) -> {
-                if (exception != null) {
-                    log.error("Failed to write state to persister for the share partition: {}-{}",
-                        groupId, topicIdPartition, exception);
-                    updatedStates.forEach(state -> state.completeStateTransition(false));
-                    future.completeExceptionally(exception);
-                    return;
-                }
+                lock.writeLock().lock();
+                try {
+                    if (exception != null) {
+                        log.error("Failed to write state to persister for the share partition: {}-{}",
+                            groupId, topicIdPartition, exception);
+                        updatedStates.forEach(state -> state.completeStateTransition(false));
+                        future.completeExceptionally(exception);
+                        return;
+                    }
 
-                log.trace("State change request successful for share partition: {}-{}",
-                    groupId, topicIdPartition);
-                updatedStates.forEach(state -> {
-                    state.completeStateTransition(true);
-                    // Cancel the acquisition lock timeout task for the state since it is acknowledged/released successfully.
-                    state.cancelAndClearAcquisitionLockTimeoutTask();
-                });
-                // Update the cached state and start and end offsets after acknowledging/releasing the acquired records.
-                maybeUpdateCachedStateAndOffsets();
-                future.complete(null);
+                    log.trace("State change request successful for share partition: {}-{}",
+                        groupId, topicIdPartition);
+                    updatedStates.forEach(state -> {
+                        state.completeStateTransition(true);
+                        // Cancel the acquisition lock timeout task for the state since it is acknowledged/released successfully.
+                        state.cancelAndClearAcquisitionLockTimeoutTask();
+                    });
+                    // Update the cached state and start and end offsets after acknowledging/releasing the acquired records.
+                    maybeUpdateCachedStateAndOffsets();
+                    future.complete(null);
+                } finally {
+                    lock.writeLock().unlock();
+                }
             });
         } finally {
             lock.writeLock().unlock();
```

Comment:
Again I would say to decouple the 2 locks and release the first one just after `updatedStates.forEach(state -> state.completeStateTransition(false));` as that only needs to be in lock.